keysign/bluetoothreceive.py

```diff
@@ -0,0 +1,144 @@
+    import keysign
+    #mod = __import__('keysign')
+    #sys.modules["keysign"] = mod
+    __package__ = str('keysign')
+
+from .gpgmh import fingerprint_from_keydata
+from .util import mac_verify
+
+log = logging.getLogger(__name__)
+
+
+class BluetoothReceive:
+    def __init__(self, port=3, size=1024):
+        self.port = port
+        self.size = size
+        self.client_socket = None
+        self.stopped = False
+
+    @inlineCallbacks
+    def find_key(self, bt_mac, mac):
+        self.client_socket = BluetoothSocket(RFCOMM)
+        message = b""
+        try:
+            self.client_socket.setblocking(False)
+            try:
+                self.client_socket.connect((bt_mac, self.port))
+            except BluetoothError as be:
+                if be.args[0] == "(115, 'Operation now in progress')":
+                    pass
+                else:
+                    raise be
+            success = False
+            while not self.stopped and not success:
+                r, w, e = yield threads.deferToThread(select.select, [self.client_socket], [], [], 0.5)
+                if r:
+                    log.info("Connection established")
+                    self.client_socket.setblocking(True)
+                    success = True
+                    # try to receive until the sender closes the connection
+                    try:
+                        while True:
+                            part_message = yield threads.deferToThread(self.client_socket.recv, self.size)
+                            message += part_message
+                    except BluetoothError as be:
+                        if be.args[0] == "(104, 'Connection reset by peer')":
+                            log.info("Bluetooth connection closed, let's check if we downloaded the key")
+                        else:
+                            raise be
+            mac_key = fingerprint_from_keydata(message)
+            verified = None
+            if mac:
+                verified = mac_verify(mac_key.encode('ascii'), message, mac)
+            if verified:
+                success = True
+            else:
+                log.info("MAC validation failed: %r", verified)
+                success = False
+                message = b""
+        except BluetoothError as be:
+            if be.args[0] == "(16, 'Device or resource busy')":
+                log.info("Probably has been provided a partial bt mac")
+            elif be.args[0] == "(111, 'Connection refused')":
+                log.info("The sender refused our connection attempt")
+            elif be.args[0] == "(112, 'Host is down')":
+                log.info("The sender's Bluetooth is not available")
+            elif be.args[0] == "(113, 'No route to host')":
+                log.info("An error occurred with Bluetooth, if present probably the device is not powered")
+            else:
+                log.info("An unknown bt error occurred: %s" % be.args[0])
```

Comment:
How about `log.exception` here?  Looks like it could make sense.
Also, the BluetoothError has a errno field which I find a bit cleaner than to match over the description of the error. But it probably works either way.  Unless those error descriptions translated.

Reply:
ah. funny. that sounds like a bug in pybluez. hrm. I would expect the Exception to set the errno instead of providing that weird string of a tuple.
Is that easily reproducible?  If so, can you cook up a small example and file a bug? Probably here: https://github.com/karulis/pybluez/issues

Reply:
It seems so, filled a bug here https://github.com/karulis/pybluez/issues/166